### PR TITLE
sd-varlink: change sd_varlink_error() to always return an error

### PR DESCRIPTION
**Pull request by poettering**
Let's make sure that sd_varlink_error() always returns an error code, so
that we can use it in a style "return sd_varlink_error(…);" everywhere,
which has two effects: return a good error reply to clients, and exit
the current stack frame with a failure code.

Interestingly sd_varlink_error_invalid_parameter() already worked like
this in some cases, but sd_varlink_error() itself didn't.

This is an alternative to the error handling tweak proposed in https://github.com/systemd/systemd/pull/34882,
but I think is a lot more generically useful, since it establishes a
pattern.

I checked our codebase, and this change should generally be OK without
breaking callsites, since the current callers (with exception of the
machined case from https://github.com/systemd/systemd/pull/34882) called sd_varlink_error() in the outermost
varlink method call dispatch stack frame, where this behaviour change
does not alter anything.

This is similar btw, how sd_bus_error_setf() and friends always return
error codes too, synthesized from its parameters.

### CODE REVIEW

File: src/libsystemd/sd-varlink/sd-varlink.c [hunk 1]
```diff
@@ -2559,7 +2566,10 @@ _public_ int sd_varlink_error(sd_varlink *v, const char *error_id, sd_json_varia
         } else
                 varlink_set_state(v, VARLINK_PROCESSED_METHOD);
 
-        return 1;
+        /* Everything worked. Let's now return the error we got passed as input as negative errno, so that
+         * programs can just do "return sd_bus_error()" and get both: a friendly error reply to clients, and
```
Comment (keszybz) [inline src/libsystemd/sd-varlink/sd-varlink.c]: Not "sd_bus_error()" :|

File: src/libsystemd/sd-varlink/sd-varlink.c [hunk 2]
```diff
@@ -1361,7 +1361,11 @@ static int varlink_dispatch_method(sd_varlink *v) {
 
                                 if (v->state == VARLINK_PROCESSING_METHOD) {
                                         r = sd_varlink_error(v, SD_VARLINK_ERROR_EXPECTED_MORE, NULL);
-                                        if (r < 0)
+                                        /* If we didn't manage to enqueue an error response, then fail the
+                                         * connection completely. Otherwise ignore the error from
+                                         * sd_varlink_error() here, it's as it is synthesized from the
+                                         * function's parameters */
```
Comment (keszybz) [inline src/libsystemd/sd-varlink/sd-varlink.c]: Missing dot.

File: src/libsystemd/sd-varlink/sd-varlink.c [hunk 3]
```diff
@@ -2558,7 +2566,10 @@ _public_ int sd_varlink_error(sd_varlink *v, const char *error_id, sd_json_varia
         } else
                 varlink_set_state(v, VARLINK_PROCESSED_METHOD);
 
-        return 1;
+        /* Everything worked. Let's now return the error we got passed as input as negative errno, so that
+         * programs can just do "return sd_bus_error()" and get both: a friendly error reply to clients, and
```
Comment (YHNdnzj) [inline src/libsystemd/sd-varlink/sd-varlink.c]: hmm? sd_bus -> varlink?

File: src/libsystemd/sd-varlink/sd-varlink.c [hunk 4]
```diff
@@ -1356,23 +1356,28 @@ static int varlink_dispatch_method(sd_varlink *v) {
...
+                                varlink_log_errno(v, r, "Method %s() called without 'more' flag, but flag needs to be set.",
                                                   method);
 
                                 if (v->state == VARLINK_PROCESSING_METHOD) {
                                         r = sd_varlink_error(v, SD_VARLINK_ERROR_EXPECTED_MORE, NULL);
-                                        if (r < 0)
-                                                return r;
+                                        /* If we didn't manage to enqueue an error response, then fail the
+                                         * connection completely. Otherwise ignore the error from
+                                         * sd_varlink_error() here, it's as it is synthesized from the
```
Comment (YHNdnzj) [inline src/libsystemd/sd-varlink/sd-varlink.c]:
```suggestion
                                         * sd_varlink_error() here, as it is synthesized from the
```

**Review by keszybz — CHANGES_REQUESTED** [review]
Looks reasonable, but I don't know varlink very well, so somebody else should take a look too.

**Review by yuwata — APPROVED** [review]
LGTM. Nice!

File: src/libsystemd/sd-varlink/sd-varlink.c [hunk 5]
```diff
@@ -2558,7 +2566,10 @@ _public_ int sd_varlink_error(sd_varlink *v, const char *error_id, sd_json_varia
         } else
                 varlink_set_state(v, VARLINK_PROCESSED_METHOD);
 
-        return 1;
+        /* Everything worked. Let's now return the error we got passed as input as negative errno, so that
+         * programs can just do "return sd_varlink_error();" and get both: a friendly error reply to clients,
+         * and an error return from the current stack frame. */
+        return sd_varlink_error_to_errno(error_id, parameters);
```
Comment (evverx) [inline src/libsystemd/sd-varlink/sd-varlink.c:2572]: Looks like at least resolved produces a lot of spurious "Invalid request descriptor" error messages on errors now
```sh
Found synthetic NXDOMAIN response.
varlink-27-27: Sending message: {"error":"io.systemd.Resolve.DNSError","parameters":{"rcode":3}}
varlink-27-27: Changing state processing-method → processed-method
Failed to send address reply: Invalid request descriptor
varlink-27-27: Connection busy.
varlink-27-27: Changing state processed-method → idle-server
varlink-27-27: Got POLLHUP from socket.
varlink-27-27: Changing state idle-server → pending-disconnect
```

Comment (yuwata) [reply src/libsystemd/sd-varlink/sd-varlink.c:2572]: Thanks. Fix is waiting in #34962.

**Pull request merged**